src/Perfmon.c

```diff
@@ -0,0 +1,192 @@
+/**@file
+ * This file is part of the TASTE Leon3 Runtime.
+ *
+ * @copyright 2022 N7 Space Sp. z o.o.
+ *
+ * TASTE Leon3 Runtime was developed under the project AURORA. 
+ * This project has received funding from the European Union’s Horizon 2020 
+ * research and innovation programme under grant agreement No 101004291
+ *
+ * Licensed under the ESA Public License (ESA-PL) Permissive,
+ * Version 2.3 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at 
+ *
+ *     https://essr.esa.int/license/list
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <Perfmon.h>
+
+
+static thread_info Perf_Mon_getThreadInfo(Perf_Mon *const perf_mon_ctx, const uint64_t threadId)
+{
+    thread_info empty_info = {};
+
+    for (int i = 0; i < RUNTIME_THREAD_COUNT; i++)
+    {
+        if(threadId == threads_info[i].id)
+        {
+            return threads_info[i];
+        }
+    }
+
+    return empty_info;
+}
+
+static bool Perf_Mon_cpuUsageVisitor(Thread_Control *the_thread, void *arg)
+{
+    Perf_Mon *const           context = (Perf_Mon*)arg;
+    char                      thread_name[MAX_THREAD_NAME_SIZE];
+    float                     usage_percent;
+    uint32_t                  integer_val;
+    uint32_t                  float_val;
+    Timestamp_Control         uptime;
+    const Timestamp_Control   used_time = _Thread_Get_CPU_time_used_after_last_reset(the_thread);
+
+    
+    _Thread_Get_name( the_thread, thread_name, sizeof(thread_name) );
+
+    _TOD_Get_uptime( &uptime );
+    _Timestamp_Subtract(&context->uptime_at_last_reset, &uptime, &context->total_usage_time);
+    _Timestamp_Divide(&used_time, &context->total_usage_time, &integer_val, &float_val);
+
+    usage_percent = (float)float_val / TOD_NANOSECONDS_PER_MICROSECOND;
+    usage_percent += (float)integer_val;
+
+    if(usage_percent < context->usage_data[context->task_iterator].cpu_usage_min_val)
+    {
+        context->usage_data[context->task_iterator].cpu_usage_min_val = usage_percent;
+    }
+
+    if(usage_percent > context->usage_data[context->task_iterator].cpu_usage_max_val)
+    {
+        context->usage_data[context->task_iterator].cpu_usage_max_val = usage_percent;
+    }
+
+    const float mean_val = context->usage_data[context->task_iterator].cpu_usage_mean_val +
+            (usage_percent - context->usage_data[context->task_iterator].cpu_usage_mean_val) /
+            (context->benchmarking_ticks + 1);
+
+    context->usage_data[context->task_iterator].cpu_usage_mean_val = mean_val;
+    context->usage_data[context->task_iterator].system_ticks_spent_by_interface_min_val =
```

Comment:
Do we need to search for the thread 3 times? Cannot we just save it to a variable?

Reply:
Done